engine-nextgen/src/main/kotlin/graphql/nadel/enginekt/instrumentation/NadelInstrumentationTimer.kt

```diff
@@ -0,0 +1,28 @@
+package graphql.nadel.enginekt.instrumentation
+
+import graphql.execution.instrumentation.InstrumentationState
+import graphql.nadel.instrumentation.NadelInstrumentation
+import graphql.nadel.instrumentation.parameters.NadelInstrumentationTimingParameters
+import graphql.nadel.instrumentation.parameters.NadelInstrumentationTimingParameters.Step
+import java.time.Duration
+
+internal class NadelInstrumentationTimer(
+    private val instrumentation: NadelInstrumentation,
+    private val userContext: Any?,
+    private val instrumentationState: InstrumentationState?,
+) {
+    inline fun <T> time(
+        step: Step,
+        function: () -> T,
+    ): T {
+        val parameters = NadelInstrumentationTimingParameters(step, userContext, instrumentationState)
+        val context = instrumentation.beginTiming(parameters)
+        val start = System.nanoTime()
+
+        return function()
```

Comment:
what if function throws an exception?

I think you should use `try catch` rather than `.also`

Reply:
Yea that's true, let me change that